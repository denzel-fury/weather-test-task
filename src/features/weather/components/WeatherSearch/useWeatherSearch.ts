import { useCallback, useState } from "react";

import debounce from "lodash.debounce";
import { useGetSuggestionsQuery } from "../../api";

interface Props {
  onSearch: (city: string) => void;
}

const useWeatherSearch = (props: Props) => {
  const { onSearch } = props;
  const [input, setInput] = useState<string>("");
  const [query, setQuery] = useState<string>("");

  const { data: suggestions } = useGetSuggestionsQuery(query, {
    skip: !query,
  });

  const debouncedSetQuery = useCallback(
    debounce((value: string) => setQuery(value), 400),
    []
  );

  const handleInput = useCallback(
    (val: string) => {
      setInput(val);
      debouncedSetQuery(val);
    },
    [debouncedSetQuery]
  );

  const handleSearch = useCallback(() => {
    onSearch(input);
  }, [onSearch, input]);

  const handleSuggestionClick = useCallback(
    (city: string) => {
      setInput(city);
      onSearch(city);
    },
    [onSearch]
  );

  return ({
    handleInput,
    handleSearch,
    handleSuggestionClick,
    suggestions,
    input,
  })
}

export default useWeatherSearch