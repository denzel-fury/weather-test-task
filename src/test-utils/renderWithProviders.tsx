import React, { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { render, RenderOptions } from '@testing-library/react';
import weatherReducer from '../features/weather/slice';

const createTestStore = () =>
  configureStore({
    reducer: {
      weather: weatherReducer,
    },
  });

const renderWithProviders = (
  ui: React.ReactElement,
  options?: RenderOptions
) => {
  const store = createTestStore();
  const Wrapper = ({ children }: PropsWithChildren) => (
    <Provider store={store}>{children}</Provider>
  );

  return render(ui, { wrapper: Wrapper, ...options });
};

export * from '@testing-library/react';
export { renderWithProviders };
