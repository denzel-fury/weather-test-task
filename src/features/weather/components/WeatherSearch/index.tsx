import React from "react";
import { useTranslation } from "react-i18next";
import useWeatherSearch from "./useWeatherSearch";
export interface WeatherSearchProps {
  onSearch: (city: string) => void;
}

const WeatherSearch: React.FC<WeatherSearchProps> = ({ onSearch }) => {
  const { t } = useTranslation();
  const {
    handleInput,
    handleSearch,
    handleSuggestionClick,
    suggestions,
    input,
  } = useWeatherSearch({ onSearch })

  return (
    <div className="flex flex-col mb-4">
      <div className="flex">
        <input
          type="text"
          value={input}
          onChange={(e) => handleInput(e.target.value)}
          placeholder={t("enterCity") ?? ""}
          className="flex-1 border p-2"
          data-testid="city-input"
        />
        <button
          onClick={handleSearch}
          className="ml-2 bg-blue-500 text-white px-4 py-2"
          data-testid="search-button"
        >
          {t("search")}
        </button>
      </div>
      {suggestions && suggestions?.length > 0 && (
        <ul className="bg-white border mt-1">
          {suggestions.map((s: string) => (
            <li
              key={s}
              className="p-2 cursor-pointer hover:bg-gray-100"
              onClick={() => handleSuggestionClick(s)}
            >
              {s}
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

export default React.memo(WeatherSearch);
