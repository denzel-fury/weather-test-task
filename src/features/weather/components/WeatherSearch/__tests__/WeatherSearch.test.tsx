import React from 'react';
import { renderWithProviders, screen, fireEvent, waitFor } from '../../../../../test-utils/renderWithProviders';
import WeatherSearch from '../index';

describe('WeatherSearch', () => {
  const mockOnSearch = jest.fn();

  beforeEach(() => {
    mockOnSearch.mockClear();
  });

  test('renders input and button', () => {
    renderWithProviders(<WeatherSearch onSearch={mockOnSearch} />);

    expect(screen.getByTestId('city-input')).toBeInTheDocument();
    expect(screen.getByTestId('search-button')).toBeInTheDocument();
  });

  test('typing into input updates its value', async () => {
    renderWithProviders(<WeatherSearch onSearch={mockOnSearch} />);
    const input = screen.getByTestId('city-input');

    fireEvent.change(input, { target: { value: 'Paris' } });

    expect((input as HTMLInputElement).value).toBe('Paris');
  });

  test('clicking search calls onSearch with input value', async () => {
    renderWithProviders(<WeatherSearch onSearch={mockOnSearch} />);
    const input = screen.getByTestId('city-input');
    const button = screen.getByTestId('search-button');

    fireEvent.change(input, { target: { value: 'Berlin' } });
    fireEvent.click(button);

    expect(mockOnSearch).toHaveBeenCalledWith('Berlin');
  });

  test('renders suggestions and handles click', async () => {
    renderWithProviders(<WeatherSearch onSearch={mockOnSearch} />);

    const input = screen.getByTestId('city-input');
    fireEvent.change(input, { target: { value: 'Lon' } });

    await waitFor(() => {
      expect(screen.getByText('London')).toBeInTheDocument();
    });

    fireEvent.click(screen.getByText('London'));

    expect(mockOnSearch).toHaveBeenCalledWith('London');
    expect((input as HTMLInputElement).value).toBe('London');
  });

  test('matches snapshot', () => {
    const { asFragment } = renderWithProviders(<WeatherSearch onSearch={() => { }} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
