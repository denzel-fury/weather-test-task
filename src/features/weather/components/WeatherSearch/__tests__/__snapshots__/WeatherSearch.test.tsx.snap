// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`WeatherSearch matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col mb-4"
  >
    <div
      class="flex"
    >
      <input
        class="flex-1 border p-2"
        data-testid="city-input"
        placeholder="enterCity"
        type="text"
        value=""
      />
      <button
        class="ml-2 bg-blue-500 text-white px-4 py-2"
        data-testid="search-button"
      >
        search
      </button>
    </div>
    <ul
      class="bg-white border mt-1"
    >
      <li
        class="p-2 cursor-pointer hover:bg-gray-100"
      >
        London
      </li>
      <li
        class="p-2 cursor-pointer hover:bg-gray-100"
      >
        Lisbon
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
